Hw1

MIZINCHIK opened the pull request.

MIZINCHIK committed:
Solve Task0

MIZINCHIK committed:
Solve Task1

MIZINCHIK committed:
Test Task1

MIZINCHIK committed:
Clean code

MIZINCHIK committed:
Fix codestyle

MIZINCHIK committed:
Fix tests for Task1

MIZINCHIK committed:
Solve Task2

MIZINCHIK committed:
Fix Task2

MIZINCHIK committed:
Fix Task2

MIZINCHIK committed:
Test Task2

MIZINCHIK committed:
Solve Task3

Solve nested arrays task using streams

MIZINCHIK committed:
Test Task3

MIZINCHIK committed:
Fix codestyle

MIZINCHIK committed:
Solve Task4

Solve fixing strings via convertion to char array

MIZINCHIK committed:
Test Task4

MIZINCHIK committed:
Solve Task5

Solve special palindrome task. Longs are converted to String straight away

MIZINCHIK committed:
Test Task5

MIZINCHIK committed:
Solve Task6

Solve Task6 via convertion to a list of digits (Integers) and sorting it in the right order.

MIZINCHIK committed:
Test Task6

MIZINCHIK committed:
Solve Task7

Solve Task7 using bitwise operations.

MIZINCHIK committed:
Test Task7

MIZINCHIK committed:
Fix codestyle

MIZINCHIK committed:
Solve Task8

Solve Task8 with a straightforward approach.

MIZINCHIK committed:
Test Task8

MIZINCHIK committed:
Fix stream usage in case of empty arrays

MIZINCHIK committed:
Fix codestyle

MIZINCHIK committed:
Fix lower limit for Task6

MIZINCHIK committed:
Fix Task6

Add zeros to make the amount of digits equal on each iterations and allow such numbers as 1000 and 9998. Add tests for edgecases.

MIZINCHIK committed:
Fix codestyle

infomouse commented on src/main/java/edu/hw1/Task1.java:
```diff
@@ -0,0 +1,26 @@
+        if (separated.length != 2 || separated[1].length() != 2) {
+            return -1;
+        }
+        long seconds = Long.parseLong(separated[1]);
```
Если в строке будет буква, то здесь получим NumberFormatException. А по условию должны возвращать -1.

Если знакомы с регулярками, то валидацию можно сделать с помощью метода String.matches().

infomouse commented on src/main/java/edu/hw1/Task8.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static boolean knightBoardCapture(int[][] chessBoard) {
+        if (chessBoard.length != BOARD_DIMENSIONS || chessBoard[0].length != BOARD_DIMENSIONS) {
```
Проверка chessBoard[0].length != BOARD_DIMENSIONS не гарантирует, что остальные строки в матрице будут такой же длины.
Пример, int[][] board = new int[][] {{1}, {2, 3}, {4, 5, 6}};

infomouse commented on src/main/java/edu/hw1/Task1.java:
```diff
@@ -0,0 +1,26 @@
+        }
+        long minutes = Long.parseLong(separated[0]);
+        boolean negative = minutes < 0;
+        seconds += Math.abs(minutes) * SECONDS_IN_MINUTE;
```
При умножении надо помнить о переполнении. minutesToSeconds(Long.MAX_VALUE + ":00") вернет -60.
Всегда надо делать тесты на граничные значения.

infomouse commented on src/main/java/edu/hw1/Task2.java:
```diff
@@ -0,0 +1,19 @@
+        throw new IllegalStateException();
+    }
+
+    public static int countDigits(double number) {
```
countDigits(Double.MIN_VALUE) возвращает 1.
Всегда надо делать тесты на граничные значения.

infomouse commented on src/main/java/edu/hw1/Task3.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public static boolean isNestable(int[] nested, int[] nest) {
+        if (nest.length == 0) {
```
В случае null будет исключение.

infomouse commented on src/main/java/edu/hw1/Task4.java:
```diff
@@ -0,0 +1,19 @@
+
+    public static String fixString(String broken) {
+        char[] characters = broken.toCharArray();
+        int strLen = characters.length;
```
Лучше не экономить буквы, stringLength намного лучше в плане читаемости кода.

infomouse commented on src/main/java/edu/hw1/Task5.java:
```diff
@@ -0,0 +1,40 @@
+        return isPalindromeDescendant(Long.toString(number));
+    }
+
+    public static boolean isPalindromeDescendant(String number) {
```
Вспомогательные методы принято делать с модификатором private.

infomouse commented on src/main/java/edu/hw1/Task5.java:
```diff
@@ -0,0 +1,40 @@
+        throw new IllegalStateException();
+    }
+
+    public static boolean isPalindromeDescendant(long number) {
```
isPalindromeDescendant(1) возвращает false, хотя это палиндром.

MIZINCHIK committed:
Fix Task1

Consider overflow cases for both negative and positive Longs. Handle NumberFormatException when parsing values that can't fit into Long. Test edgecases.

MIZINCHIK committed:
Fix codestyle

MIZINCHIK committed:
Fix array size checks in Task8 using stream

MIZINCHIK committed:
Fix Task2 switching to integer numbers

As we are counting the number of digits in a number and not its representation but the floating-point numbers do not have that property the solution was switched to Long instead.

MIZINCHIK committed:
Check on null arrays in Task3. Test edgecases

MIZINCHIK committed:
Fix naming in Task4

MIZINCHIK committed:
Fix Task5

Fix acces modifiers, single-digit case, delete tests for private methods.

MIZINCHIK committed:
Revert namings in Task5 to required

infomouse commented on src/main/java/edu/hw1/Task2.java:
```diff
@@ -0,0 +1,19 @@
+        throw new IllegalStateException();
+    }
+
+    public static int countDigits(long number) {
```
Ошибка: countDigits(Long.MIN_VALUE) возвращает 1.
Всегда надо делать тесты на граничные значения.

MIZINCHIK committed:
Fix handling negative numbers in Task2 & Add tests for edgecases
